Node WorkerChannel native addon

jmillan opened the pull request:
Substitute spawning a process with mediasoup for running mediasoup in a thread created by this addon.

The communication is now done via message passing rather than using unixsocket, avoiding the kernel networking stack for each and every message passed from Node to C++ and viceversa.

Performance gains are enhanced by approximately 8 times. Tested by sending a request from Node to C++ and waiting for the response N times in a loop.

Caveats:

* Since we don't use a separate process, a crash in mediasoup would crash Node, meaning it must be handled by node process signal handling.
  * At the same time, the coredumps will be named something-**node**-something
  * In order to debug coredumps the worker-channel.node binary can be used along with the coredump file and gdb/lldb 

TODO:

- [ ] Document removed 'died' and 'subprocessclose' Worker.ts events.
- [ ] Document removed 'died()' and 'subprocessClosed()' Worker.ts methods.
- [ ] Missing all the prebuilt binary upload/download stuff.

jmillan committed:
WIP: Node WorkerChannel native addon

Substitute spawning a process with mediasoup for running mediasoup in a
thread created by this addon.

The communication is now done via message passing rather than using
unixsocket, avoiding the kernel networking stack for each and every
message passed from Node to C++ and viceversa.

Performance gains are enhanced by approximately 8 times. Tested by
sending a request from Node to C++ and waiting for the response N times
in a loop.

jmillan committed:
remove worker-channel from dependencies in package.json

ibc commented on .gitignore:
```diff
@@ -10,6 +10,12 @@
+## Node WorkerChannel addon.
+/node/workerChannel/node_modules
+/node/workerChannel/build
+/node/workerChannel/lib
```
Shouldn't we include this `lib` folder in "files" in main `package.json`? Or should it be managed by the `package.json` of the addon folder? What I mean is that:

- We only include JS transpiled files in the mediasoup NPM package and not TS files (AKA `node/src` folder).
- Is the addon properly built if we install mediasoup as a git+ssh dependency in the package.json of whatever app using mediasoup?

ibc replied:
Reminder

jmillan replied:
Done and tested. package.json `files` are properly edited. adding mediasoup as a git+ssh dependency of a 3rd party app has also been tested.

ibc commented:
`.prettierignore` and `.eslintignore` should include `/node/workerChannel/lib`.

ibc commented on node/src/Worker.ts:
```diff
@@ -203,7 +202,6 @@ export type WorkerDump = {
 
 export type WorkerEvents = {
 	died: [Error];
-	subprocessclose: [];
```
This is a documented event, why is it being removed? I mean, we still may need to know when the native addon and its worker thread are closed, am I wrong?

jmillan replied:
They are never closed by themselves. The only way to close them is via `worker.close()` method in worker.ts, which sends a `WORKER_CLOSE` request to the worker. That should release all the libuv resources. Once that happens the thread running `mediasoup_worker_run()` will close, and when the JS `worker` instance is garbage collected, the native instance will be destroyed.

jmillan replied:
I still have to do more testing on the closure part BTW.

jmillan replied:
Added to the TODO list in the description.

ibc commented on node/src/Worker.ts:
```diff
@@ -247,11 +245,13 @@ const workerLogger = new Logger('Worker');
 
 	// Worker process PID.
-	readonly #pid: number;
+	readonly #pid: number = process.pid;
```
This amazes me. Does it mean that we only use the CPU core in which the Node process runs for **all the workers**? Ok, in theory those are native threads so they may use different CPU cores, is it?

jmillan replied:
Correct, threads are not pinned to specific CPU.

ibc commented on node/src/Worker.ts:
```diff
@@ -521,13 +398,6 @@ export class Worker<
 		return this.#died;
 	}
 
-	/**
-	 * Whether the Worker subprocess is closed.
-	 */
-	get subprocessClosed(): boolean {
-		return this.#subprocessClosed;
-	}
```
This requires changes in the public docs.

jmillan replied:
Item added to the TODO list.

ibc commented on node/workerChannel/src/workerChannel.hpp:
Can we please organize public and private methods and member fields the same way we do in any other C++ class?

jmillan replied:
no problem

ibc commented on node/workerChannel/src/workerChannel.hpp:
```diff
@@ -0,0 +1,34 @@
+// mediasoup header files.
+#include "common.hpp"
+#include "lib.hpp"
+
```
Remove blank line please, we don't do this anywhere.

ibc commented on npm-scripts.mjs:
```diff
@@ -128,6 +129,12 @@ async function run() {
 			break;
 		}
 
+		case 'worker:build-lib': {
+			buildWorkerLib();
+
+			break;
+		}
```
Is there any real reason to add this as a public script in `package.json`? I removed many of these in the past because it didn't make any sense to make them public. Does it make sense that we call `npm run worker:build-lib` directly? And why is this about "worker build lib"?

jmillan replied:
leftover.

ibc commented on npm-scripts.mjs:
```diff
@@ -315,6 +322,20 @@ function buildWorker() {
+	const buildType = process.env.MEDIASOUP_BUILDTYPE || 'Release';
+
+	executeCmd(
+		`cd node/workerChannel && GYP_DEFINES="mediasoup_build_type=${buildType}" npm i`
```
Can we use `npm ci` instead please? And copy `.npmrc` from root folder to `node/workerChannel`.

jmillan replied:
Also, the prebuilt stuff is missing totally. We will need to upload the library now, not the binary.

ibc commented on package.json:
```diff
@@ -54,6 +54,7 @@
 		"typescript:build": "node npm-scripts.mjs typescript:build",
 		"typescript:watch": "node npm-scripts.mjs typescript:watch",
 		"worker:build": "node npm-scripts.mjs worker:build",
+		"worker:build-lib": "node npm-scripts.mjs worker:build-lib",
```
As expected, this is not needed at all, please remove.

jmillan replied:
I've been using it while development, left over. Remember this is still a DRAFT.

jmillan committed:
provide version

jmillan committed:
add workerChannels to .eslintignore and .prettierignore

jmillan committed:
move header file to include folder

jmillan committed:
remove blank line

jmillan committed:
package.json: remove worker:build-lib script

jmillan committed:
install workerChannel via 'npm ci'

jmillan committed:
properly build Debug/Release build types

jmillan committed:
lint

ibc commented:
Please merge v3 and check prettier and eslint path arrays in npm-scripts.mjs

jmillan committed:
remove .eslintignore and .prettierignore

jmillan committed:
Merge remote-tracking branch 'origin/v3' into node-worker-channel

jmillan committed:
add workerChannel to eslint and prettier

jmillan committed:
fix name for libmediasoup-worker library in linux

jmillan committed:
add *.ts files inside src/ folder

jmillan committed:
workerChannel: npm-scripts.ts

jmillan committed:
eslint

jmillan committed:
npm-scripts: build worker lib before typescript

jmillan committed:
npm-scripts: remove typo

jmillan committed:
workerChannel: initialize pointer

And use std::lock_guard for better RAII

jmillan committed:
binding.gyp: use a common path for libmediasoup-worker library

jmillan committed:
binding.gyp: use module_root_dir rather than pwd

Which should work in windows too.

jmillan committed:
Windows: test

jmillan committed:
npm-scripts: do not build mediasoup-worker but libmediasoup-worker

jmillan committed:
remove worker died event

jmillan committed:
windows test

jmillan committed:
windows test cont.

jmillan committed:
windows test

jmillan committed:
workflow remove every OS but windows

jmillan committed:
Windows: test

jmillan committed:
binding.gyp: remove unneeded dependency

jmillan committed:
worker-channel: remove unused dependency

jmillan committed:
npm-scripts: build addon with --verbose option

jmillan committed:
meson.build: set b_vscrt option to use /MT runtime library for windows

jmillan committed:
windows: test

jmillan committed:
windows: test, add more needed libraries for windows

ibc commented on node/workerChannel/npm-scripts.mjs:
lint and prettier missing, or better include this file in main npm-scripts.

jmillan replied:
Do you mean not having an `npm-scripts` in node/workerChannel and do all related actions: build/lint/prettier, etc from the main `npm-scripts` file?

ibc replied:
Probably it makes sense. It's part of the Node layer, it's not an optional one such as worker/scripts/ which is only needed for worker development.

jmillan replied:
Lint and prettier are already done in main npm-scripts as I've added the given files path there, and in .eslintrc. But for building I think It's also nice to be able to build the addon in isolation, hence having a npm-scripts file there makes sense.

For example, you are changing the `WorkerChannel.cpp` and run `npm ci` in `node/workerChannel` which does not rebuild libmediasoup-worker but just rebuilds the addon.

What do you think?

ibc replied:
ok

ibc commented on node/workerChannel/src/index.ts:
```diff
@@ -0,0 +1,36 @@
+	`../build/${buildType}/worker-channel.node`
+);
+
+export class WorkerChannel extends EventEmitter {
```
This should use our EnhancedEventEmitter with proper types.

ibc commented on node/workerChannel/src/workerChannel.cpp:
```diff
@@ -0,0 +1,197 @@
+#include "../include/workerChannel.hpp"
+#include "napi.h"
+
```
Remove blank line.

ibc commented on node/workerChannel/src/workerChannel.cpp:
```diff
@@ -0,0 +1,197 @@
+	for (uint32_t i = 0; i < params.Length(); i++)
+	{
+		Napi::Value v = params[i];
+		if (!v.IsString())
```
Blank line.

ibc commented on node/workerChannel/src/workerChannel.cpp:
```diff
@@ -0,0 +1,197 @@
+
+WorkerChannel::~WorkerChannel()
+{
+	std::cout << "~WorkerChannel()" << std::endl;
```
????

jmillan replied:
I need to do some tests with garbage collector, this log it temporal.

jmillan committed:
workerChannel: include uv.h rather than DepLibUV.hpp

In order to avoid the error:

error LNK2005: uv_async_send already defined in node.lib(node.exe)

jmillan committed:
windows: WIP, do not import libuv

jmillan committed:
handle feedback

jmillan committed:
tests

jmillan committed:
add missing include

jmillan committed:
windows: linker option to allow multiple symbols

Temporal work around until libmediasoup-worker does not expose libuv
symbols

jmillan committed:
format

jmillan committed:
use EnhancedEventEmitter rather than EventEmitter

jmillan committed:
be able to provide a specific libmediasoup_worker.a library

jmillan committed:
binding.gyp: use single quotes

jmillan committed:
Windows: only add debug symbols on Debug build type

jmillan committed:
npm-scripts: use a cons for addon path

jmillan committed:
package.json: fix name and description

jmillan committed:
Revert "Windows: only add debug symbols on Debug build type"

This reverts commit 76426445b08526a8188f02501290f3625cfaa316.

debug library is needed to create minidump files when the process
crashes.

jmillan committed:
make workerChannel a completely separate npm module

jmillan committed:
make workerChannel a completely separate npm module

jmillan committed:
add missing Logger and enhancedEvents files

jmillan committed:
Revert "make workerChannel a completely separate npm module"

This reverts commit 4237fc26fd1a899fb750120e521ffea375d80406.

jmillan committed:
do not use a separate TS or npm project for WorkerChannel

jmillan committed:
lint

jmillan committed:
npm-scripts: test, --forceExit

Node will clean the native addon once it garbage collects the
corresponding JS instance.

ibc commented on npm-scripts.mjs:
```diff
@@ -441,7 +459,7 @@ function flatcWorker() {
 	logInfo('testNode()');
 
-	executeCmd(`jest --silent false --detectOpenHandles ${args}`);
+	executeCmd(`jest --silent false --detectOpenHandles --forceExit ${args}`);
```
Please don't

jmillan replied:
Garbage collection on addons does not happen at our will. The native addon will be closed once the JS instance is garbage collected.

Within the tests, the addon is not garbage collected so the tests won't finish unless this option is set.

jmillan replied:
Anyway I'm rechecking this because I need to see how it is garbage collected.

jmillan replied:
Handled. It turns out that I need to set the JS instance to `undefined` to tell the GC that it is not longer used so it can clean it up, which will perform the destruction of the C++ class. I mean, if I don't do it then I assume GC would sometime in the future do it anyway but that would mean that the tests would not finish due to the open handles.

ibc replied:
While it makes sense honestly I don't think that Node triggers garbage collection IMMEDIATELY as soon as an object is unreferenced. It may take some time depending on Node internal things and, if so, we will get random warnings here.

jmillan replied:
Correct, it does not trigger GC at all. Setting a variable to  `undefined` just let the GC identify that the object is not used.

In this exact case, now the tests do finish while before they didn't unless the `--forceExit` option was set. This wouldn't be needed for real life, but we don't want the test to hang (CI would fail) nor use `--forceExit` since it could shadow real problems.

jmillan committed:
workerChannel: add into node/src

jmillan committed:
WorkerChannel: .gitignore

jmillan committed:
workflow: enable all SO

jmillan committed:
lint

jmillan committed:
workflow: fix tabs

jmillan committed:
clang-format

jmillan committed:
clang-format: typo

jmillan committed:
test: remove --forceExit

assign JS instance of the addon to 'undefined' when closed, to indicate
the Node GC that it can be freed.

Otherwise the tests won't finish due to the correponsing handles being
opened.

jmillan committed:
.gitignore

jmillan committed:
worker: format

jmillan committed:
windows: set RuntimeLibrary for Debug and Release

jmillan committed:
workflow: enable all OS in mediasoup-node

jmillan committed:
tests: add comment of addon tests

jmillan committed:
WIP: prebuild

jmillan committed:
lint

jmillan committed:
WIP: prebuilt

jmillan committed:
WIP: prebuilt

jmillan committed:
WIP: prebuilt

jmillan committed:
prebuilt working

jmillan committed:
comment

jmillan committed:
workerChannel: Do not copy the addon binary into node/lib

jmillan committed:
package.json: files, add workerChannel files

jmillan committed:
.gitignore

jmillan committed:
cleanup

jmillan committed:
package.json: engines, set a minimum npm version

Lower npm versions than 10.3.0 fail building the addon.

Error: No module named 'distutils'

jmillan commented:
Update:

This PR is almost ready to be shipped if we want to.

The are two main concerns already defined in the PR description:

* Since we don't use a separate process for the application and the worker, a crash in mediasoup would crash Node, meaning it must be handled by node process signal handling. This is true for Rust too.
  *  The coredumps will be named something-node-something
  *  In order to debug coredumps the worker-channel.node binary can be used along with the coredump file and gdb/lldb
* Since the worker is now used in multi-thread fashion, the very same concerns apply to Node as to Rust, such us https://github.com/versatica/mediasoup/issues/1352

Having said this, do we still want to go in this direction?. As indicated in the description:

> Performance gains are enhanced by approximately 8 times. Tested by sending a request from Node to C++ and waiting for the response N times in a loop.

But of course it comes with the mentioned caveats too.

What do you think @ibc, @nazar-pc?

nazar-pc commented:
If there use cases where the performance gap is that large, I think it is worth it. The proper solution to crashes is largely to write code that doesn't crash rather than running it in a separate process.

ibc commented:
I think that keeping the current design just to keep the "benefit" of not crashing the node process if the worker fails is not a good argument so I am ok but:

How does this change affect prebuilt worker binaries? Now they should not be "binaries" but "libraries" so more changes are needed specially in the naming of those assets and the function in npm-script that fetches them.

jmillan commented:
> How does this change affect prebuilt worker binaries? Now they should not be "binaries" but "libraries" so more changes are needed specially in the naming of those assets and the function in npm-script that fetches them.

Yes, there is a TODO for this.

:+1:,

jmillan committed:
Merge remote-tracking branch 'origin/v3' into node-worker-channel

jmillan committed:
remove unneeded 'async' keyword

jmillan committed:
Remove unused variables

jmillan committed:
typo

jmillan committed:
package-lock.json

jmillan committed:
cosmetic

jmillan committed:
update tests

jmillan committed:
aREADME

jmillan committed:
fix test

jmillan committed:
update package-lock.json

jmillan committed:
rename .hpp and .cpp files for consistency

jmillan committed:
rename files, cont.

jmillan committed:
add .clang-format

jmillan committed:
lint

jmillan committed:
update gyp file to use newer MACOSX_DEPLOYMENT_TARGET

jmillan commented:
I've updated the branch with 'v3'. Current status is that there are some linking issues when using `MESON_ARGS=-Db_sanitize=thread`. I guess the same options need to be provided to the addon.

jmillan committed:
test: sanitizer

jmillan committed:
test: sanitizer, link against libasan

In order to avoid having to compile node with sanitizer enabled

jmillan committed:
fix typo

jmillan committed:
test linking asan and tsan dinamically

jmillan committed:
test, cont

jmillan committed:
TMP: reduce CI matrix

jmillan committed:
Revert "TMP: reduce CI matrix"

This reverts commit 5a887dd88a7032d9edee0cc7886086cc192a9814.

jmillan committed:
remove sanitization from node tests

We are not even checking the CI logs.

Worker sanitization should be done at worker level.

jmillan committed:
erase node-gyp configure

jmillan committed:
gyp. cleanup

jmillan committed:
fix binding.gyp

jmillan committed:
TMP: only run windows CI

jmillan committed:
adapt windows runtime library

jmillan committed:
reduce CI instances

jmillan committed:
Disable windows CI for the time being

jmillan commented:
@nazar-pc, this PR is up to date with v3, but I've commented out Windows usage on node CI actions because it [fails](https://github.com/versatica/mediasoup/actions/runs/18317003142/job/52159932229) when building the add-on. 

The .gyp file for the add-on is [this](https://github.com/versatica/mediasoup/blob/549d9b2e4508426f3f378ddceee1038729f345c4/node/src/workerChannel/binding.gyp). I'm quite sure it used to work before (months ago).

Since you have some other times fixed some other Windows issues, I wonder if you could take a look whenever you find some free slots.

jmillan committed:
binding.gyp, remove usnused line

jmillan committed:
remove no-exceptions flag globally

Only set them for macos and linux

jmillan committed:
CI reenable windows

nazar-pc commented:
Looks like gyp can't find visual studio. Maybe it doesn't support the latest MSVC version?

Specifically this line is concerning:
> subprocess.CalledProcessError: Command '['cl', '-dM', '-E', '-x', 'c', 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpnwi1n4h7.c']' returned non-zero exit status 1.

Grok suggested to try this (yes, `cmd` instead of PowerShell) to set it up explicitly:
```
- name: Install dependencies
  shell: cmd
  run: |
    call "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64
    npm install
```

You might have to set up a Windows VM to debug what is going on there rather than pushing countless test commits here.

jmillan commented:
> You might have to set up a Windows VM to debug what is going on there rather than pushing countless test commits here

This is my dream 😊

ibc commented:
> Grok suggested to try this

OMG the Nazi AI of the Nazi Elon Musk 🥸
